### PR TITLE
Scrubbing LogisticRegression learners

### LINKED ISSUE #2615: Scrubbing LogisticRegression learners
Subtask for #2613 
`LogisticRegression`
`MulticlassLogisticRegression`
`PoissonRegression`
and their base classes

### PR DESCRIPTION
**Pull request by abgoswam**
Fixes #2615.  Related #2613

### CODE REVIEW

**Comment by Ivanidzo4ka** [comment]
>             public float L2Weight = Defaults.L2Weight;

does L2Regularization sound better for you?

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:28 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>             public int MemorySize = Defaults.MemorySize;

I'm looking on this one, and I think it would be amount of memory be allocated in MB during training.
But in reality it's 
```
/// <summary>
/// Number of previous iterations to remember for estimate of Hessian.
/// </summary>
```

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:45 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public abstract class OptionsBase : LearnerInputBaseWithWeight
             [Argument(ArgumentType.AtMostOnce, HelpText = "Maximum iterations.", ShortName = "maxiter")]
             [TGUI(Label = "Max Number of Iterations")]
             [TlcModule.SweepableLongParamAttribute("MaxIterations", 1, int.MaxValue)]
-            public int MaxIterations = Defaults.MaxIterations;
+            public int MaximumIterations = Defaults.MaximumIterations;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >MaximumIterations [](start = 23, length = 17)

let's stick to NumberOfIterations across all our learners.

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 2]
```diff
@@ -71,7 +71,7 @@ public abstract class OptionsBase : LearnerInputBaseWithWeight
             [Argument(ArgumentType.LastOccurenceWins, HelpText = "Init weights diameter", ShortName = "initwts", SortOrder = 140)]
             [TGUI(Label = "Initial Weights Scale", SuggestedSweeps = "0,0.1,0.5,1")]
             [TlcModule.SweepableFloatParamAttribute("InitWtsDiameter", 0.0f, 1.0f, numSteps: 5)]
-            public float InitWtsDiameter = 0;
+            public float InitialWeightsDiameter = 0;
 
             // Deprecated
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >// Deprecated [](start = 12, length = 13)

:)
can you double check it's true?

**Comment by Ivanidzo4ka** [comment]
>             public float L1Weight = Defaults.L1Weight;

L1Regularization ?

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:33 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>             public bool ShowTrainingStats = false;

you change it to ShowTrainingStatistics  in LogisticRegression

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/MulticlassLogisticRegression.cs:49 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         public static LogisticRegression LogisticRegression(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,

I think we trying to have following pattern for trainer names:
`{AlgoName}{PredictionKind}Trainer`
like above we have
`SdcaBinaryTrainer` (not a great example tho. it should be `StochasticDualCoordinateAscentBinaryTrainer`) or LightGBMMulticlassTrainer.

Actually created issue regarding that: [https://github.com/dotnet/machinelearning/issues/2762](https://github.com/dotnet/machinelearning/issues/2762) stay tuned!



---
Refers to: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs:427 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

**Comment by abgoswam** [comment]
>             public int MemorySize = Defaults.MemorySize;

i will add that as comment .. 

in some places its also commented as follows without reference to the Hessian:

/// <param name="m">The number of previous iterations to store</param>



---
In reply to: [467967467](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467967467) [](ancestors = 467967467)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:45 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>             public int MemorySize = Defaults.MemorySize;

I would rather change parameter name to reflect what it does than update documentation.

---
In reply to: [467984224](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467984224) [](ancestors = 467984224,467967467)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:45 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 3]
```diff
@@ -71,7 +71,7 @@ public abstract class OptionsBase : LearnerInputBaseWithWeight
             [Argument(ArgumentType.LastOccurenceWins, HelpText = "Init weights diameter", ShortName = "initwts", SortOrder = 140)]
             [TGUI(Label = "Initial Weights Scale", SuggestedSweeps = "0,0.1,0.5,1")]
             [TlcModule.SweepableFloatParamAttribute("InitWtsDiameter", 0.0f, 1.0f, numSteps: 5)]
-            public float InitWtsDiameter = 0;
+            public float InitialWeightsDiameter = 0;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >InitialWeightsDiameter [](start = 25, length = 22)

So when you do this, you need to take old name and put it in ShortName in ArgumentAttribute like this ShortName="initwts,initwtsDiameter"

Comment (abgoswam) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: 1. does it apply to all the other renamings as well ?  .
2. why  we adding it to ShortName like that ?  

---
In reply to: [260893400](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260893400) [](ancestors = 260893400)

**Comment by abgoswam** [comment]
>             public float L1Weight = Defaults.L1Weight;

fixed

---
In reply to: [467969539](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467969539) [](ancestors = 467969539)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:33 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 1 above)
Comment (abgoswam) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: fixed

---
In reply to: [260873873](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260873873) [](ancestors = 260873873)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 2 above)
Comment (abgoswam) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: looks like its used...  one of our tests `MulticlassLogisticRegressionOnnxConversionTest` even uses it by setting it to `false`

i have added a comment

---
In reply to: [260874466](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260874466) [](ancestors = 260874466)

**Comment by abgoswam** [comment]
>             public float L2Weight = Defaults.L2Weight;

fixed

---
In reply to: [467966407](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467966407) [](ancestors = 467966407)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:28 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: My question is it actually used in logisticRegression code?
Across all other learners we just use `NumberOfThreads` to pass how many threads we want to use.
I just don't see reason why this one should be remain public.

---
In reply to: [260904491](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260904491) [](ancestors = 260904491,260874466)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 3 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: 1. Yes
2. for cmd back compatibility.

---
In reply to: [260894964](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260894964) [](ancestors = 260894964,260893400)

**Comment by abgoswam** [comment]
>         public static LogisticRegression LogisticRegression(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,

marking as pending . maybe outside purview of this PR

we should have separate PR for #2762 once names are finalized

---
In reply to: [467981004](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467981004) [](ancestors = 467981004)

---
Refers to: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs:427 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs (hunk 2 above)
Comment (abgoswam) [reply src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: Made internal for now.  

I see several usage of this in TestEntryPoint.cs   .  As you pointed, we should look into whether we can remove this completely  and only use `NumberOfThreads` even from EntryPoints




---
In reply to: [260905721](https://github.com/dotnet/machinelearning/pull/2761#discussion_r260905721) [](ancestors = 260905721,260904491,260874466)

**Comment by Ivanidzo4ka** [comment]
>         protected virtual void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contributions, int top, int bottom, bool normalize)

should be private protected

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LinearModelParameters.cs:265 in 04a4f7a. [](commit_id = 04a4f7a367aed75978dad0d0b0cb1e1d8ea4301f, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>             public int MemorySize = Defaults.MemorySize;

Feels a bit too long.
IterationsToRemember, IterationsHistory, IterationsToKeep?

---
In reply to: [467985202](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-467985202) [](ancestors = 467985202,467984224,467967467)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:45 in 9ea2076. [](commit_id = 9ea2076862712cb9856ecaade204d658cf2061f4, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 4]
```diff
@@ -22,33 +22,51 @@ public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : Trainer
...
             [TlcModule.SweepableLongParamAttribute("MaxIterations", 1, int.MaxValue)]
-            public int MaxIterations = Defaults.MaxIterations;
+            public int NumberOfIterations = Defaults.NumberOfIterations;
 
+            /// <summary>
+            /// Run SGD to initialize LR weights, converging to this tolerance.
+            /// </summary>
             [Argument(ArgumentType.AtMostOnce, HelpText = "Run SGD to initialize LR weights, converging to this tolerance",
                 ShortName = "sgd")]
             public float SgdInitializationTolerance = 0;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >SgdInitializationTolerance [](start = 25, length = 26)

StochasticGradientDescentInitilaizationTolerance ?

**Comment by Ivanidzo4ka** [comment]
>                         Console.Write(".");

Oh god, It still exist.
this need to be ported to progress channels.

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:337 in 04a4f7a. [](commit_id = 04a4f7a367aed75978dad0d0b0cb1e1d8ea4301f, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>                         Console.Write(".");

Not part of your PR, so please ignore.

---
In reply to: [468851424](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-468851424) [](ancestors = 468851424)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:337 in 04a4f7a. [](commit_id = 04a4f7a367aed75978dad0d0b0cb1e1d8ea4301f, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs [hunk 5]
```diff
@@ -191,31 +215,31 @@ internal static class Defaults
...
-            Host.CheckParam(!(LbfgsTrainerOptions.MemorySize <= 0), nameof(LbfgsTrainerOptions.MemorySize), "Must be positive, if provided.");
-
-            L2Weight = LbfgsTrainerOptions.L2Weight;
-            L1Weight = LbfgsTrainerOptions.L1Weight;
-            OptTol = LbfgsTrainerOptions.OptTol;
-            MemorySize =LbfgsTrainerOptions.MemorySize;
-            MaxIterations = LbfgsTrainerOptions.MaxIterations;
-            SgdInitializationTolerance = LbfgsTrainerOptions.SgdInitializationTolerance;
+            Host.CheckUserArg(!LbfgsTrainerOptions.UseThreads || LbfgsTrainerOptions.NumberOfThreads > 0 || LbfgsTrainerOptions.NumberOfThreads == null,
+              nameof(LbfgsTrainerOptions.NumberOfThreads), "numThreads must be positive (or empty for default)");
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs]: >numThreads [](start = 60, length = 10)

nameof

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/MulticlassLogisticRegression.cs [hunk 6]
```diff
@@ -45,8 +45,11 @@ public sealed class MulticlassLogisticRegression : LbfgsTrainerBase<MulticlassLo
 
         public sealed class Options : OptionsBase
         {
-            [Argument(ArgumentType.AtMostOnce, HelpText = "Show statistics of training examples.", ShortName = "stat", SortOrder = 50)]
-            public bool ShowTrainingStats = false;
+            /// <summary>
+            /// If set to <value>true</value>training statistics will be generated at the end of training.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Show statistics of training examples.", ShortName = "stat, ShowTrainingStats", SortOrder = 50)]
+            public bool ShowTrainingStatistics = false;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/MulticlassLogisticRegression.cs]: >ShowTrainingStatistics [](start = 24, length = 22)

Can you double check it's used?
It strange binary LR and MultiLR has different options (binary LR has extra thing `ComputeLogisticRegressionStandardDeviation`)

**Comment by artidoro** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

Should we rewrite this without acronyms? 

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Comment by abgoswam** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

There is a separate issue #2762  from trainer names..maybe we can address changes (if any) as part of  #2762

the scrubbing PRs are focussed on the items mentioned in #2613  

---
In reply to: [469360727](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469360727) [](ancestors = 469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Comment by artidoro** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

I  see "no short names" in the items mentioned in #2613. I thought it was part of the scrubbing tasks?

---
In reply to: [469362212](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469362212) [](ancestors = 469362212,469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

File: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/MulticlassLogisticRegression.cs [hunk 7]
```diff
@@ -45,8 +45,11 @@ public sealed class MulticlassLogisticRegression : LbfgsTrainerBase<MulticlassLo
 
         public sealed class Options : OptionsBase
         {
-            [Argument(ArgumentType.AtMostOnce, HelpText = "Show statistics of training examples.", ShortName = "stat", SortOrder = 50)]
-            public bool ShowTrainingStats = false;
+            /// <summary>
+            /// If set to <value>true</value>training statistics will be generated at the end of training.
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/MulticlassLogisticRegression.cs]: Could you add a space before training?

**Comment by abgoswam** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

please note :  "no short names" is for Options , not trainer names


---
In reply to: [469364309](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469364309) [](ancestors = 469364309,469362212,469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Comment by artidoro** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

OK I see!

---
In reply to: [469376815](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469376815) [](ancestors = 469376815,469364309,469362212,469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Comment by artidoro** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

Sorry for this then

---
In reply to: [469377354](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469377354) [](ancestors = 469377354,469376815,469364309,469362212,469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Comment by abgoswam** [comment]
>     public abstract class LbfgsTrainerBase<TOptions, TTransformer, TModel> : TrainerEstimatorBase<TTransformer, TModel>

no worries at all .. thanks for noticing actually  :)

---
In reply to: [469377514](https://github.com/dotnet/machinelearning/pull/2761#issuecomment-469377514) [](ancestors = 469377514,469377354,469376815,469364309,469362212,469360727)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsPredictorBase.cs:18 in 7a7ac78. [](commit_id = 7a7ac7852be6b29629731e18bce5ea69af5f7c3b, deletion_comment = False)

**Pull request merged**